File: ios/Ninehoto/Models/SessionStatistics.swift
import Foundation

struct SessionStatistics: Codable, Equatable {
    var swipeCount: Int
    var deleteCount: Int
    var keepCount: Int
    var sessionDate: Date
    var durationSeconds: TimeInterval
    var averageSwipesPerMinute: Double
    var spaceSavedBytes: Int64

    static var empty: SessionStatistics {
        SessionStatistics(
            swipeCount: 0,
            deleteCount: 0,
            keepCount: 0,
            sessionDate: Date(),
            durationSeconds: 0,
            averageSwipesPerMinute: 0,
            spaceSavedBytes: 0
        )
    }

    var deleteRate: Double {
        guard swipeCount > 0 else { return 0 }
        return Double(deleteCount) / Double(swipeCount) * 100
    }

    var keepRate: Double {
        guard swipeCount > 0 else { return 0 }
        return Double(keepCount) / Double(swipeCount) * 100
    }

    var formattedDuration: String {
        let minutes = Int(durationSeconds) / 60
        let seconds = Int(durationSeconds) % 60
        if minutes > 0 {
            return "\(minutes)m \(seconds)s"
        }
        return "\(seconds)s"
    }
}

final class StatisticsTracker {
    static let shared = StatisticsTracker()

    private let userDefaults = UserDefaults.standard
    private let keyPrefix = "statistics_"

    private init() {}

    func saveSession(_ statistics: SessionStatistics) {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        let key = keyPrefix + dateFormatter.string(from: statistics.sessionDate)

        if let data = try? JSONEncoder().encode(statistics) {
            userDefaults.set(data, forKey: key)
        }

        PreferencesManager.shared.lastSessionDate = statistics.sessionDate
    }

    func getAllSessions() -> [SessionStatistics] {
        var sessions: [SessionStatistics] = []

        for key in userDefaults.dictionaryRepresentation().keys {
            guard key.hasPrefix(keyPrefix),
                  let data = userDefaults.data(forKey: key),
                  let statistics = try? JSONDecoder().decode(SessionStatistics.self, from: data) else {
                continue
            }
            sessions.append(statistics)
        }

        return sessions.sorted { $0.sessionDate > $1.sessionDate }
    }

    func getTotalStatistics() -> SessionStatistics {
        let sessions = getAllSessions()

        let totalSwipes = sessions.reduce(0) { $0 + $1.swipeCount }
        let totalDeletes = sessions.reduce(0) { $0 + $1.deleteCount }
        let totalKeeps = sessions.reduce(0) { $0 + $1.keepCount }
        let totalDuration = sessions.reduce(0) { $0 + $1.durationSeconds }
        let totalSpace = sessions.reduce(0 as Int64) { $0 + $1.spaceSavedBytes }

        return SessionStatistics(
            swipeCount: totalSwipes,
            deleteCount: totalDeletes,
            keepCount: totalKeeps,
            sessionDate: Date(),
            durationSeconds: totalDuration,
            averageSwipesPerMinute: totalSwipes > 0 ? (totalSwipes / (totalDuration / 60)) : 0,
            spaceSavedBytes: totalSpace
        )
    }

    func clearAll() {
        for key in userDefaults.dictionaryRepresentation().keys {
            if key.hasPrefix(keyPrefix) {
                userDefaults.removeObject(forKey: key)
            }
        }
    }
}

File: android/Ninehoto.Android/Models/SessionStatistics.cs
using System;
using System.Collections.Generic;

namespace Ninehoto.Android.Models
{
    public class SessionStatistics
    {
        public int SwipeCount { get; set; }
        public int DeleteCount { get; set; }
        public int KeepCount { get; set; }
        public DateTime SessionDate { get; set; }
        public double DurationSeconds { get; set; }
        public double AverageSwipesPerMinute { get; set; }
        public long SpaceSavedBytes { get; set; }

        public static SessionStatistics Empty => new SessionStatistics
        {
            SwipeCount = 0,
            DeleteCount = 0,
            KeepCount = 0,
            SessionDate = DateTime.Now,
            DurationSeconds = 0,
            AverageSwipesPerMinute = 0,
            SpaceSavedBytes = 0
        };

        public double DeleteRate => SwipeCount > 0 ? (double)DeleteCount / SwipeCount * 100 : 0;
        public double KeepRate => SwipeCount > 0 ? (double)KeepCount / SwipeCount * 100 : 0;

        public string FormattedDuration
        {
            get
            {
                var minutes = (int)DurationSeconds / 60;
                var seconds = (int)DurationSeconds % 60;
                return minutes > 0 ? $"{minutes}m {seconds}s" : $"{seconds}s";
            }
        }
    }

    public class StatisticsTracker
    {
        private const string KeyPrefix = "statistics_";
        private readonly Android.Content.ISharedPreferences _prefs;

        public StatisticsTracker(Android.Content.ISharedPreferences prefs)
        {
            _prefs = prefs;
        }

        public void SaveSession(SessionStatistics statistics)
        {
            var key = KeyPrefix + statistics.SessionDate.ToString("yyyy-MM-dd");
            var json = Newtonsoft.Json.JsonConvert.SerializeObject(statistics);
            _prefs.Edit().PutString(key, json).Apply();
        }

        public List<SessionStatistics> GetAllSessions()
        {
            var sessions = new List<SessionStatistics>();
            var allEntries = _prefs.All;

            foreach (var entry in allEntries)
            {
                if (entry.Key.StartsWith(KeyPrefix) && entry.Value is string json)
                {
                    try
                    {
                        var stats = Newtonsoft.Json.JsonConvert.DeserializeObject<SessionStatistics>(json);
                        if (stats != null) sessions.Add(stats);
                    }
                    catch { }
                }
            }

            sessions.Sort((a, b) => b.SessionDate.CompareTo(a.SessionDate));
            return sessions;
        }

        public SessionStatistics GetTotalStatistics()
        {
            var sessions = GetAllSessions();
            var total = SessionStatistics.Empty;

            foreach (var session in sessions)
            {
                total.SwipeCount += session.SwipeCount;
                total.DeleteCount += session.DeleteCount;
                total.KeepCount += session.KeepCount;
                total.DurationSeconds += session.DurationSeconds;
                total.SpaceSavedBytes += session.SpaceSavedBytes;
            }

            total.AverageSwipesPerMinute = total.DurationSeconds > 0 
                ? total.SwipeCount / (total.DurationSeconds / 60) 
                : 0;

            return total;
        }

        public void ClearAll()
        {
            var editor = _prefs.Edit();
            var keys = new List<string>();
            
            foreach (var entry in _prefs.All)
            {
                if (entry.Key.StartsWith(KeyPrefix))
                    keys.Add(entry.Key);
            }

            foreach (var key in keys)
                editor.Remove(key);
            
            editor.Apply();
        }
    }
}